Mixpanel/MixpanelExceptionHandler.m

```diff
@@ -93,88 +108,66 @@ - (void)addMixpanelInstance:(Mixpanel *)instance {
-                                          exceptionWithName:UncaughtExceptionHandlerSignalExceptionName
-                                          reason:
-                                          [NSString stringWithFormat:
-                                           NSLocalizedString(@"Signal %d was raised.", nil),
-                                           signal]
-                                          userInfo:
-                                          [NSDictionary
-                                           dictionaryWithObject:[NSNumber numberWithInt:signal]
-                                           forKey:UncaughtExceptionHandlerSignalKey]] waitUntilDone:YES];
+    struct sigaction prev_action = handler.prev_signal_handlers[signal];
+    if (prev_action.sa_flags & SA_SIGINFO) {
+        if (prev_action.sa_sigaction) {
+            prev_action.sa_sigaction(signal, info, context);
+        }
+    } else if (prev_action.sa_handler) {
+        prev_action.sa_handler(signal);
+    }
 }
 
-void MPHandleException(NSException *exception)
-{
+void MPHandleException(NSException *exception) {
+    MixpanelExceptionHandler *handler = [MixpanelExceptionHandler sharedHandler];
+
     int32_t exceptionCount = OSAtomicIncrement32(&UncaughtExceptionCount);
-    if (exceptionCount > UncaughtExceptionMaximum)
-    {
-        return;
+    if (exceptionCount <= UncaughtExceptionMaximum) {
+        NSArray *callStack = [MixpanelExceptionHandler backtrace];
+        NSMutableDictionary *userInfo = [[exception userInfo] mutableCopy];
+        userInfo[UncaughtExceptionHandlerAddressesKey] = callStack;
+        NSException *exceptionWithCallstack = [NSException exceptionWithName:[exception name]
+                                                                      reason:[exception reason]
+                                                                    userInfo:userInfo];
+
+        [handler mp_handleUncaughtException:exceptionWithCallstack];
     }
 
-    NSArray *callStack = [MixpanelExceptionHandler backtrace];
-    NSMutableDictionary *userInfo =
-    [NSMutableDictionary dictionaryWithDictionary:[exception userInfo]];
-    [userInfo
-     setObject:callStack
-     forKey:UncaughtExceptionHandlerAddressesKey];
-
-    [MixpanelExceptionHandler performSelectorOnMainThread:@selector(mp_handleUncaughtException:)
-                              withObject:[NSException
-                                          exceptionWithName:[exception name]
-                                          reason:[exception reason]
-                                          userInfo:userInfo] waitUntilDone:YES];
+    if (handler.defaultExceptionHandler) {
+        handler.defaultExceptionHandler(exception);
+    }
 }
 
 
-+ (void) mp_handleUncaughtException:(NSException *)exception {
-    MixpanelExceptionHandler *handler = [MixpanelExceptionHandler sharedHandler];
-
+- (void) mp_handleUncaughtException:(NSException *)exception {
```

Comment:
afaik the main thread is the last to die and if we dont do this on the main thread then we might not do it at all as there is a race condition.

Reply:
i thought it would continue until the function returned. specifically, for a while when i was running something on the main thread i think there was a deadlock and the app hung forever when crashing (i think only from signals). not sure if that was because crashlytics was handling on the main thread as well or what, but anyway it seemed safe to do from the crashing thread. i don't think PLCrashReporter did it on the main thread, they have a comment saying "The callback will execute on the crashed thread." i can look into it more if you want